[quant][graphmode][refactor] Change QParamMap to QParamVector

jerryzh168 opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #33650 [not4land] debugging
* #33533 [not4land] Temporary workaround to get quantized ops
* #34855 [quant][graphmode] quantization support for prim::CallFunction
* #34346 [quant][graphmode] Add quantization support for aten::cat
* #34345 [quant][graphmode] Add prim::ListConstruct to general op handling
* #34807 [quant][graphmode] quantization support for prim::ListUnpack
* #34854 [quant][graphmode] quantization support for view, transpose, contiguos
* #34806 [quant][graphmode] quantization support for aten::chunk
* #34805 [quant][graphmode] add constant prop in finalize to simplify graph
* #34804 [quant][graphmode] Replicate quantize node for prim::If
* #34572 [quant][graphmode] quantization support for aten::add
* #34518 [quant][graphmode] quantization work for prim::If
* #34571 [quant][graphmode][fix] use observed_values_ to check values are observed
* #34414 [quant][graphmode][refactor] insertObservers for Block
* #34411 [quant][graphmode][fix] Insert dequantize before use node
* #34349 [quant][graphmode][fix] preserve the type of original value when inserting dequant node
* #34348 [quant] Add dequantize.tensors
* #34803 [quant][graphmode] quantization support for aten::rehshape
* #34347 [quant][graphmode] Add quantization support for aten::dropout
* #34315 [quant][graphmode] insert quant/dequant work for duplicated debugName
* **#34314 [quant][graphmode][refactor] Change QParamMap to QParamVector**
* #33854 [quant][graphmode] Quantization pattern for aten::linear

Summary:

Test Plan:
.

Reviewers:
.

Subscribers:

Tasks:

Tags:

Differential Revision: [D20493032](https://our.internmc.facebook.com/intern/diff/D20493032)

jerryzh168 committed:
[quant][graphmode][refactor] Change QParamMap to QParamVector

Summary:

Test Plan:
.

Reviewers:
.

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

jerryzh168 committed:
Update on "[quant][graphmode][refactor] Change QParamMap to QParamVector"


Summary:

Test Plan:
.

Reviewers:
.

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

jerryzh168 committed:
Update on "[quant][graphmode][refactor] Change QParamMap to QParamVector"


Summary:

Test Plan:
.

Reviewers:
.

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

jerryzh168 committed:
Update on "[quant][graphmode][refactor] Change QParamMap to QParamVector"


Summary:

Test Plan:
.

Reviewers:
.

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

raghuramank100 commented on torch/csrc/jit/passes/quantization.cpp:
```diff
@@ -30,7 +30,7 @@ using NameModuleVector = std::vector<std::pair<std::string, script::Module>>;
 // for example _scale, _zero_point,
 // _scalar_type and _axis(for per channel quantization)
-using QParamMap = std::unordered_map<std::string, IValue>;
+using QParamVector = std::vector<std::pair<std::string, IValue>>;
```
Can you point to a test that checks per channel quantization?

jerryzh168 replied:
https://github.com/pytorch/pytorch/blob/master/test/test_jit.py#L1291

jerryzh168 committed:
Update on "[quant][graphmode][refactor] Change QParamMap to QParamVector"


Summary:

Test Plan:
.

Reviewers:
.

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

jerryzh168 committed:
Update on "[quant][graphmode][refactor] Change QParamMap to QParamVector"


Summary:

Test Plan:
.

Reviewers:
.

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

jerryzh168 committed:
Update on "[quant][graphmode][refactor] Change QParamMap to QParamVector"


Summary:

Test Plan:
.

Reviewers:
.

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

jerryzh168 committed:
Update on "[quant][graphmode][refactor] Change QParamMap to QParamVector"


Summary:

Test Plan:
.

Reviewers:
.

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

jerryzh168 committed:
Update on "[quant][graphmode][refactor] Change QParamMap to QParamVector"


Summary:

Test Plan:
.

Reviewers:
.

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

jerryzh168 committed:
Update on "[quant][graphmode][refactor] Change QParamMap to QParamVector"


Summary:

Test Plan:
.

Reviewers:
.

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

jerryzh168 committed:
Update on "[quant][graphmode][refactor] Change QParamMap to QParamVector"


Summary:

Test Plan:
.

Reviewers:
.

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

supriyar commented on torch/csrc/jit/passes/quantization.cpp:
```diff
@@ -1376,19 +1381,19 @@ std::tuple<c10::QScheme, QParamMap> InsertQuantDeQuantHelper::
-    qparams["_scale"] = scale;
-    qparams["_zero_point"] = zero_point;
-    qparams["_axis"] = tp->elements()[2].toInt();
+  if (isPerChannel(qscheme)) {
```
Should we be checking if `qscheme` is valid before this line?

jerryzh168 replied:
I think we don't have invalid qscheme right now: https://bddppq.github.io/codebrowser/pytorch/pytorch/c10/core/QScheme.h.html

supriyar reviewed (approved):
One comment, otherwise looks good.

jerryzh168 committed:
Update on "[quant][graphmode][refactor] Change QParamMap to QParamVector"


Summary:

Test Plan:
.

Reviewers:
.

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

jerryzh168 committed:
Update on "[quant][graphmode][refactor] Change QParamMap to QParamVector"


Summary:

Test Plan:
.

Reviewers:
.

Subscribers:

Tasks:

Tags:

Differential Revision: [D20493032](https://our.internmc.facebook.com/intern/diff/D20493032)

[ghstack-poisoned]
